Move Nav APIs to new platform

joshdover opened the pull request:
## Summary

Subpart of #18843 and #34091

This moves the core Nav APIs from `ui/chrome` into the `ChromeService` in the new platform.

- `ChromeStart` now exposes a sub-service for reading and making limited updates to navlinks. These are powered by apps registered with the `ApplicationService` and filtered by UI Capabilities before being exposed by the `ChromeService`.
- The `header-global-nav` directive now consumes navlinks from the new platform.
- The `lastSubUrl` feature utilized by legacy apps has been refactored and will remain in `ui/chrome`. This feature utilizes the limited fields that `ChromeService` exposes to updates by outside code.

This change is the main blocker to moving the Chrome UI to the new platform. This will be necessary to enable the new platform to control top-level routing.

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- [ ] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)
- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios
- [ ] This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)

### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)
- [ ] This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8101/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8113/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8337/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8902/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8939/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9075/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9113/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9848/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9879/)

joshdover committed:
Move chrome navlinks to core

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11887/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12980/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12994/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13012/)

elasticmachine commented:
Pinging @elastic/kibana-platform

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13037/)

jgowdyelastic commented on src/core/public/chrome/nav_links/nav_link.ts:
```diff
@@ -0,0 +1,114 @@
+  /**
+   * Disables a link from being clickable.
+   *
+   * NOTE: this is only used by the ML and Graph plugins currently. They use this field
```
This comment is not technically true from the point of view of ML.
If the user's platinum license has expired, ML will still be available in the nav bar but some features will be disabled within the ML app, such as creating new jobs.

The disabling of the ML icon in the nav bar originally happened if we could not load the x-pack license information. This appears to be broken, and has been broken before this PR.
We originally copied this behaviour from another x-pack app, Graph i believe, but it was a long time ago.

jgowdyelastic reviewed (approved):
Code changes for ML LGTM.
At some point in the future we might want to look at what should be available to plugins when adding icons to the nav bar.
It appears the ability to disable an icon is broken or not available any more. This has been the case from before this PR. Is it a feature that is needed at all?
Looking at the license checking in most plugins, they are all quite similar, probably due to people copying code. When we first came to add ML to kibana we copied the way an existing plugin was doing it. 
Because of this I believe there is some redundant code which people have added to their plugins because they assumed they needed to. 
For example, most plugins add an `enableLinks` flag to the features passed to the client side.
It looks like logstash and reporting are making use of this flag, but others aren't.

joshdover committed:
Update generated API docs

eliperelman commented on src/legacy/ui/public/chrome/directives/header_global_nav/components/header.tsx:
```diff
@@ -235,31 +238,27 @@ class HeaderUI extends Component<Props, State> {
+        href: addBasePath(navLink.href),
+        iconType: navLink.euiIconType,
+        icon:
+          !navLink.euiIconType && navLink.icon ? (
```
You can remove the explicit undefined with:

```js
icon: !navLink.euiIconType && navLink.icon && (
  <EuiImage
    size="s"
    alt=""
    aria-hidden={true}
    url={chrome.addBasePath(`/${navLink.icon}`)}
  />
),
```

Of course, that assumes that `icon` will be `undefined` and not some other falsy value, so if that's the case, please  leave this as is.

eliperelman commented on x-pack/plugins/ml/public/hacks/toggle_app_link_in_nav.js:
```diff
@@ -7,19 +7,20 @@
-  if (!chrome.navLinkExists('ml')) return;
 
-  const navLink = chrome.getNavLinkById('ml');
+  const navLinkUpdates = {};
```
Could this object be simplified to:

```js
const showAppLink = xpackInfo.get('features.ml.showLinks', false);
const navLinkUpdates = {
  hidden: !showAppLink,
  disabled: showAppLink && !xpackInfo.get('features.ml.isAvailable', false),
};
```

eliperelman reviewed (changes requested):
Looks good, just have a couple questions.

joshdover commented on src/legacy/ui/public/chrome/api/nav.ts:
```diff
@@ -0,0 +1,163 @@
+  function refreshLastUrl(link: ChromeNavLink) {
+    const lastSubUrl = getLastUrl(link);
+
+    return coreNavLinks.update(link.id, {
```
The interface only allows a few fields to be modified, and it does updates by creating new objects rather than modifying the existing ones. The fields that are allowed to be modified are currently only being enforced by TypeScript, I'll change that to be enforced at runtime as well and make sure these objects are frozen.

eliperelman replied:
Is this meant to be intentionally mutable? Have we considered any issues that may arise from mutating external objects, or is that the intent?

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13130/)

joshdover commented on src/core/public/chrome/nav_links/nav_links_service.ts:
```diff
@@ -0,0 +1,144 @@
+  private readonly stop$ = new ReplaySubject(1);
+
+  public start({ application }: StartDeps) {
+    const navLinks$ = new BehaviorSubject<ReadonlyArray<NavLinkWrapper>>(
```
I think that makes sense as well, though a `Map` will make more sense than a `Set` since these are keyed by `id`.

eliperelman replied:
I feel like this service could benefit from storing the nav links in a Set rather than an array. This would simplify the implementation of `get`, `exists`, and others. In fact, if this service could mimic the operations of Set, that would be nice, e.g. rename `exists` to `has`. Thoughts?

joshdover commented on src/legacy/core_plugins/tests_bundle/tests_entry_template.js:
```diff
@@ -94,7 +94,8 @@ const coreSystem = new CoreSystem({
         user: {}
-      }
+      },
+      nav: []
```
This code isn't linted because it's actually a big template string, but I'll add go ahead and add the trailing comma.

eliperelman replied:
Not sure why this isn't linted, but I think this should have a trailing comma.

joshdover commented on src/legacy/ui/public/chrome/api/__tests__/nav.js:
```diff
@@ -79,94 +71,119 @@ describe('chrome nav apis', function () {
-      } = init({ appUrlStore, nav });
+      } = init({
+        appUrlStore,
+        nav: [
```
This is enforced by prettier which is not applied to js files. I'll add go ahead and add the trailing commas.

eliperelman replied:
Good catch, and thanks for the info!

eliperelman replied:
These do not have trailing commas either, so we may want to check the linter settings to see if there are other problems.

joshdover commented on src/legacy/ui/public/chrome/api/nav.ts:
```diff
@@ -0,0 +1,163 @@
+  internals.trackPossibleSubUrl = async function(url: string) {
+    const kibanaParsedUrl = absoluteToParsedUrl(url, chrome.getBasePath());
+
+    for (let link of coreNavLinks.getAll()) {
```
I think I actually prefer using a mutable `link` variable here so that you can't introduce a bug by accidentally not using the most recent version of `link`.

On `forEach` vs. `for ... of` I'm a bit split. I've started leaning towards `for ... of` simply because you cannot use async/await with `forEach`.

eliperelman replied:
> On `forEach` vs. `for ... of` I'm a bit split. I've started leaning towards `for ... of` simply because you cannot use async/await with `forEach`.

Just for posterity, this is solved by using `map` instead of `forEach` with `Promise.all`. Use `map` when you want "parallel" async actions, and `for/of` when you want "serial" async actions:

```js
// "parallel"
await Promise.all(
  items.map(async item => {
    await something(item);
  });
);

// "serial"
for (const item of items) {
  await something(item);
}
```

eliperelman replied:
If you wanted this to be less imperative and avoid mutable variables, this could be written as:

```js
coreNavLinks.getAll().forEach(link => {
  const subUrlBase = legacyNavProps.get(link.id)!.subUrlBase;
  const active = kibanaParsedUrl.getAppRootPath().startsWith(subUrlBase);
  const updatedLink = coreNavLinks.update(link.id, { active })!;

  if (active) {
    setLastUrl(link, url);
    return;
  }

  const refreshedUrlLink = refreshLastUrl(link);
  const newGlobalState = kibanaParsedUrl.getGlobalState();

  if (newGlobalState) {
    injectNewGlobalState(link, kibanaParsedUrl.appId, newGlobalState);
  }
});
```

Not a requirement, just an option. :smile:

joshdover committed:
Review comments

joshdover commented:
> At some point in the future we might want to look at what should be available to plugins when adding icons to the nav bar.
> It appears the ability to disable an icon is broken or not available any more. This has been the case from before this PR. Is it a feature that is needed at all?
> Looking at the license checking in most plugins, they are all quite similar, probably due to people copying code. When we first came to add ML to kibana we copied the way an existing plugin was doing it.
> Because of this I believe there is some redundant code which people have added to their plugins because they assumed they needed to.
> For example, most plugins add an `enableLinks` flag to the features passed to the client side.
> It looks like logstash and reporting are making use of this flag, but others aren't.

I'm going to dig into why this is broken exactly, it may be that the Nav UI code is not reading the `disabled` property correctly.

My idea for this going forward is to allow applications to update their status via the ApplicationService in the new platform, rather than interfacing with Chrome APIs to update links. This way, the ApplicationService can manage displaying a license page of information or redirecting to somewhere else in Kibana.

eliperelman reviewed (approved):
Looks great!

joshdover committed:
Merge remote-tracking branch 'upstream/master' into chrome-navlinks

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13150/)

joshdover committed:
Small tweaks

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13187/)

eliperelman commented on src/core/public/chrome/chrome_service.mock.ts:
```diff
@@ -53,17 +54,36 @@ const createSetupContractMock = () => {
   return setupContract;
 };
 
+const createStartContractMock = (): jest.Mocked<ChromeStart> => {
```
nit: you can remove the redundant curly braces and return:

```js
const createStartContractMock = (): jest.Mocked<ChromeStart> => ({
  navLinks: {
    getNavLinks$: jest.fn(),
    clear: jest.fn(),
    exists: jest.fn(),
    get: jest.fn(),
    getAll: jest.fn(),
    showOnly: jest.fn(),
    update: jest.fn(),
    enableForcedAppSwitcherNavigation: jest.fn(),
    getForceAppSwitcherNavigation$: jest.fn(),
  },
});
```

eliperelman commented on src/core/public/chrome/nav_links/nav_links_service.ts:
```diff
@@ -0,0 +1,144 @@
+       */
+      get(id: string) {
+        const link = navLinks$.value.find(l => l.id === id);
+        return link ? link.properties : undefined;
```
If you want to avoid the explicit return of undefined and the ternary, you could write as:

```js
return link && link.properties;
```

eliperelman commented on src/legacy/ui/public/chrome/api/nav.ts:
```diff
@@ -0,0 +1,163 @@
+      const includedId = deletedIds.find(deletedId => {
+        return url.includes(deletedId);
+      });
+      if (includedId === undefined) {
```
This can be simplified to:

```js
return includedId !== undefined;
```

spalger commented on src/legacy/ui/public/chrome/directives/header_global_nav/components/header.tsx:
```diff
@@ -88,11 +91,11 @@ const TRUNCATE_LIMIT: number = 64;
-  const href = relativeToAbsolute(chrome.addBasePath(recentlyAccessed.link));
-  const navLink = navLinks.find(nl => href.startsWith(nl.subUrlBase));
+  const href = chrome.addBasePath(recentlyAccessed.link);
+  const navLink = navLinks.find(nl => href.startsWith(nl.appUrl));
```
This no longer finds the navLink for recently accessed items when basePath is in use, because `nl.appUrl` doesn't have the basePath, but `href` does.

spalger commented on src/legacy/ui/public/chrome/directives/header_global_nav/components/header.tsx:
```diff
@@ -235,31 +238,28 @@ class HeaderUI extends Component<Props, State> {
+      .map(navLink => ({
+        key: navLink.id,
+        label: navLink.title,
+        href: addBasePath(navLink.href),
```
This is necessary now because extendNavLink is no longer setting the actual href value but including the `appUrl` sometimes (which doesn't have the basePath) and absolute urls the rest of the time. I feel like we probably don't need both of these.

spalger reviewed (commented):
Things are looking really good, but I'm pretty sure that subUrl handling is applying outside of Kibana and I'm pretty sure that wasn't the case before recently (maybe before this change)

joshdover committed:
Merge remote-tracking branch 'upstream/master' into chrome-navlinks

joshdover committed:
Use absolute URLs to handle multiple Kibana instances on a domain

joshdover commented:
@spalger I've updated this to fix the bug when there are multiple Kibana instances running under different base paths on the same domain. This requires using mostly absolute URLs as the internal representation (as it was before this change). I've also renamed `appUrl` (which wasn't actually a URL) to `baseUrl` (which now is an absolute URL).

spalger commented on x-pack/plugins/ml/public/hacks/toggle_app_link_in_nav.js:
```diff
@@ -7,19 +7,19 @@
-    navLink.disabled = !xpackInfo.get('features.ml.isAvailable', false);
-  }
+
+  const navLinkUpdates = {
```
👍 want to apply the same treatment to `x-pack/plugins/graph/public/hacks/toggle_app_link_in_nav.js`?

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13428/)

joshdover committed:
Fix recently accessed + lint

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13439/)

joshdover committed:
Fix icon mapping for recentlyAccessed

Bring subUrlBase into new platform ☹️

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13455/)

spalger reviewed (approved):
LGTM!
